DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp

```diff
@@ -0,0 +1,124 @@
+#ifndef __CACHED_LOCAL_MEMORY_HPP__
+#define __CACHED_LOCAL_MEMORY_HPP__
+
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+
+#include "constexpr_math.hpp"   // DirectProgramming/DPC++FPGA/include
+
+namespace fpga_tools {
+
+template <typename T,             // type to store in the memory
+          size_t k_mem_depth,     // depth of the memory
+          size_t k_cache_depth    // number of elements in the cache
+         >
+class CachedLocalMemory {
+ public:
+  static constexpr int kNumAddrBits = fpga_tools::CeilLog2(k_mem_depth);
+  using addr_t = ac_int<kNumAddrBits, false>;
+
+  CachedLocalMemory() {
+    for (int i = 0; i < k_cache_depth; i++) {
+      cache_valid_[i] = false;
+    }
+  }
+  CachedLocalMemory(T init_val) {
```

Comment:
What is the point of having an initializer if all the valids are false?

Reply:
There is only a valid per cache entry.  If all the cache entries are invalid, then a read will pull data from the actual memory.  If the data in the actual memory has not been initialized, then you read garbage.

Reply:
Oh right, sorry for some reason I was thinking that the initialization was for the cache values! My bad!